### ISSUE TITLE
Usability suggestions -- Atom, bilayer

### ISSUE THREAD

**Issue by zerothi** [zerothi/sisl#713]
**Describe the feature**

Would it make sense that:

```python
C, B, N = Atom("C, B, N")
bilayer = geom.bilayer(..., atoms=[C, (B, N)])
```
I believe the first one is simple, and the other one is more of semantics to make it look like the rest of the interfaces. The idea being that we can provide top/bottom atoms in a single go.

What is the opinion here?

@ialcon @tfrederiksen @pfebrer

**Comment by ialcon** [comment]
I understand that within `bilayer` you'll have other options specifying, for e.g., the type of bilayer to be created (e.g. hexagonal layers), right?

Also, what I've seen so far regarding atomic categories is that these are used as a means to identify atoms with particular properties within an input structure (e.g. all C atoms, all atoms with 3 bonds, and so on). However, you are using here the category `Atom()` apparently as an input to the `bilayer` function - as a means to tell it I want a bilayer formed by a graphene flake with an hBN one, right? This just strikes me as a rather different usage from the one I though categories were meant to...

**Comment by zerothi** [comment]
> I understand that within `bilayer` you'll have other options specifying, for e.g., the type of bilayer to be created (e.g. hexagonal layers), right?

The bilayer creates hexagonal structures of bilayers (bilayer graphene), see [here](https://zerothi.github.io/sisl/api/generated/sisl.geom.bilayer.html#bilayer).

Agreed, it should probably be renamed to honeycomb_bilayer for consistency :)

> 
> Also, what I've seen so far regarding atomic categories is that these are used as a means to identify atoms with particular properties within an input structure (e.g. all C atoms, all atoms with 3 bonds, and so on). However, you are using here the category `Atom()` apparently as an input to the `bilayer` function - as a means to tell it I want a bilayer formed by a graphene flake with an hBN one, right? This just strikes me as a rather different usage from the one I though categories were meant to...

This is not about categories, this is about geometry creation (not related to categories at all). It is whether it is more intuitive to do:
```python
ChBN = geom.bilayer(..., atoms=[C, (B, N)])
# or
ChBN = geom.bilayer(..., bottom_atoms=C, top_atoms=(B, N))
```
which is the intuitive way of using the bilayer construct?

**Comment by ialcon** [comment]
Ok, I got confused with the categories terminology... Well, this is really a stylistic thing, right? Using `bottom_atoms` and `top_atoms` is nice because is clearer what one is specifying, but of course something like `atoms=[C, (B, N)]` is more compact, and after some thinking the user will also understand what you are doing there... I kind of like both... sorry! :stuck_out_tongue:

**Comment by zerothi** [comment]
Hehe, thanks! :)

**Comment by pfebrer** [comment]
This:

```
C, B, N = Atom("C, B, N")
```

I think it's not a good idea to use strings for multiple values. Why not just pass a list?

For the other one I have no preference, only a comment on some value that might be ambiguous:

```
ChBN = geom.bilayer(..., atoms=C)
```

means all atoms are carbon, right? So what would

```
ChBN = geom.bilayer(..., atoms=(B, N))
```

mean? Is it to create two layers with B and N or one layer with B and another with N?

**Comment by zerothi** [comment]
> This:
> 
> ```python
> C, B, N = Atom("C, B, N")
> ```
> 
> I think it's not a good idea to use the same string for multiple values. Why not just pass a list?
Ok, I actually agree. The above is just as easy as passing `"C B N".split()`, we won't do it.
> 
> For the other one I have no preference, only a comment on some value that might be ambiguous:
> 
> ```python
> ChBN = geom.bilayer(..., atoms=C)
> ```
> 
> means all atoms are carbon, right? So what would
> 
> ```python
> ChBN = geom.bilayer(..., atoms=(B, N))
> ```
> 
> mean? Is it to create two layers with B and N or one layer with B and another with N?

Yes, as you say. If one wants everything alternating, one should do `((B, N), (B, N))`. It was more to stream-line the interfaces, but it could be superfluous in this case?

**Comment by ialcon** [comment]
From what I understand in your example, @zerothi, atoms of different layers are specified in a list given to the `atoms` key of the `geom.bilayer` function, right?
```python
ChBN = geom.bilayer(..., atoms=[C, (B, N)])
```
if this is the case, then specifying anything not being a list, like in both these cases:
```python
ChBN = geom.bilayer(..., atoms=C)
ChBN = geom.bilayer(..., atoms=(B, N))
```
would apply to both layers, and one should provide a list (with 2 elements) to specify (2) layers of different composition. Does that make sense?

**Comment by ialcon** [comment]
Well, I guess it just depends on how you want to organize the specification of layer composition in the `atoms` key, no? If you want to use a tupple so that:
```python
ChBN = geom.bilayer(..., atoms=(B, N))
```
means a B-top-layer and a N-bottom-layer, then specifying a bilayer composed of graphene and hBN should be something like:
```python
ChBN = geom.bilayer(..., atoms=(C, [B, N]))
```
or:
```python
ChBN = geom.bilayer(..., atoms=(C, (B, N)))
```
no?

**Comment by ialcon** [comment]
But of course, if you prefer to leave it as it is, go ahead! It will be a functional notation, I believe, in either case

**Comment by zerothi** [comment]
Yes, so the question is, *are these usages intuitive*? If not, we should leave it. :)

**Comment by ialcon** [comment]
I think they are. It might be worth to enforce having to use a tupple and so to specify two elements, even if they are the same. So that bilayer graphene would be:
```python
biGr = geom.bilayer(..., atoms=(C, C))
```
This, actually, is a shorter notation than specifying only once the element (of this homogeneous bilayer) via something like `bottom_atoms=C`, and you kind of gain the feeling that a bilayer is composed of two things (two layers) whose compositions *need* to be specified. From here, I think that it is clear that if you have a heterogeneous layer, one of the two elements in this tupple will necessarily be a list, or another tupple (I kind of like more `[B, N]` for hBN than `(B, N)`, just to avoid confusion with the "bilayer" tupple - but I think this is a minor thing, and quite subjective anyways...)

**Comment by zerothi** [comment]
@pfebrer @tfrederiksen do you have anything to add? Is it *intuitive* to remove the `bottom|top_atoms` and unify with only `atoms`?

**Comment by pfebrer** [comment]
Hmm it can be unified into `atoms`, but then I would add some examples in the docstring.

**Comment by zerothi** [comment]
It seems like this might be a bit confusion, I will close this issue, and then we can always return if it makes sense. Thanks for all the input!
